Add functions to use symmetric encryption with provided IVs [MAID-2918]

nbaksalyar opened the pull request:
* Add functions to use symmetric encryption with custom initialisation vectors. We use a randomly generated unique nonce as an IV by default.
* Also add a function to create a symmetric encryption key out of a byte array.

nbaksalyar committed:
feat/symmetric: add functions to use symmetric encryption with provided IVs

MAID-2918

povilasb commented on src/lib.rs:
```diff
@@ -71,8 +71,18 @@ use serde::{de::DeserializeOwned, Serialize};
+pub type PublicSignKey = [u8; PUBLIC_SIGN_KEY_SIZE];
+
+/// Initialisation vector.
+pub type Iv = [u8; IV_SIZE];
```
Maybe have a longer name for a public type? e.g. `InitVec`, `InitializationVec`, etc.

povilasb commented on src/lib.rs:
```diff
@@ -352,6 +362,14 @@ impl SymmetricKey {
         }
     }
 
+    /// Create a new key from bytes.
```
Just curious, in what cases do we store our encryption key serialized? :)

nbaksalyar replied:
@povilasb, the access container in the Authenticator actually stores symmetric encryption keys in the serialised form (to be more exact, it stores the [`MDataInfo`](https://github.com/maidsafe/safe_client_libs/blob/4f8b40f53fad1ff9b874085dd10c51ad0090c42a/safe_core/src/client/mdata_info.rs) structs that can point to private containers, along with their encryption key).

povilasb commented on src/lib.rs:
```diff
@@ -364,6 +382,17 @@ impl SymmetricKey {
+    /// Encrypts bytestring `plaintext` using initialisation vector `iv`.
+    ///
+    /// Returns ciphertext in case of success.
+    pub fn encrypt_bytes_iv(&self, plaintext: &[u8], iv: Iv) -> Result<Vec<u8>, Error> {
```
Even though we provide custom nonce, does the requirement of using it only once still apply? If so, I think we should explicitly state that to make it harder to misuse this function.

nbaksalyar replied:
Great catch, thanks!

However, in future versions we'll replace the symmetric encryption algo with Miscreant (which is [_nonce-misuse resistant_](https://github.com/miscreant/miscreant/wiki/Nonce-Reuse-Misuse-Resistance)) and this won't apply in that case.

povilasb replied:
Ciphers with nonce reuse misuse resistance, such as the AES-SIV and AES-PMAC-SIV ciphers provided by this library, do not fail catastrophically in this case: the only information these ciphers leak is that if the same message is encrypted with the same nonce twice under the same key, we obtain the same ciphertext. **Though this information is still potentially useful to attackers and thus we should still avoid reusing nonces if possible,**

Seems like we still should discourage `safe_crypto` users from reusing nonces.

povilasb commented on src/lib.rs:
```diff
@@ -364,6 +382,17 @@ impl SymmetricKey {
+    ///
+    /// Returns ciphertext in case of success.
+    pub fn encrypt_bytes_iv(&self, plaintext: &[u8], iv: Iv) -> Result<Vec<u8>, Error> {
+        Ok(secretbox::seal(
```
If we can never fail, why return `Error`?

nbaksalyar committed:
refactor/symmetric: use better naming

Address review points:
1. Rename the `Iv` type to `Nonce`.
2. Rename `encrypt_bytes_iv` to `encrypt_bytes_with_nonce` (same for decryption function).
3. Do not return `Error` for `encrypt_bytes_with_nonce`.

nbaksalyar committed:
feat/nonce: add a nonce generation function

ustulation commented on src/lib.rs:
```diff
@@ -92,6 +102,11 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
 }
 
+/// Generate a new unique nonce.
+pub fn gen_nonce() -> Nonce {
```
what's the use of this - is it not misleading when you already have things like [these](https://github.com/maidsafe/safe_crypto/pull/15/files#diff-b4aea3e418ccdb71239b96952d9cddb6L288) ?
